sparse_ho/models.py

```diff
@@ -196,15 +195,14 @@ def _update_only_jac_sparse(
     def _reduce_alpha(alpha, mask):
         return alpha
 
-    # @staticmethod
-    def _get_jac_t_v(self, jac, mask, dense, alphas, v):
-        n_samples = self.X.shape[0]
+    @staticmethod
+    def _get_jac_t_v(X, y, jac, mask, dense, alphas, v, n_samples):
         return n_samples * alphas[mask] * np.sign(dense) @ jac
 
-    def proj_param(self, log_alpha):
+    def proj_param(self, X, y, log_alpha):
```

Comment:
I don't get what this does

Reply:
As before, it ensure that the parameters ```log_alpha``` does not take value absurdly large, typically we bound it between ```log_alpha_max``` and ```log_alpha_max - 8```

Reply:
proj regularizer ? clip regularizer ? This is costly to recompute alpha_max every time, couldn't we pass max_val, min_val and use np.clip ?

Reply:
I went for ```proj_hyperparam```